src/test/path-mapping.spec.ts

```diff
@@ -0,0 +1,132 @@
+import { join } from 'path';
+
+import { createExec } from './exec-helpers';
+import {
+  CMD_ESM_LOADER_WITHOUT_PROJECT,
+  CMD_TS_NODE_WITHOUT_PROJECT_FLAG,
+  nodeSupportsEsmHooks,
+  TEST_DIR,
+  installTsNode,
+} from './helpers';
+import { test } from './testlib';
+
+test.beforeAll(installTsNode);
+
+const execBuilder = (
+  command: string,
+  moduleDir: string,
+  tsConfig: string = 'tsconfig.json'
+) => {
+  const partialExec = createExec({
+    cwd: join(TEST_DIR, moduleDir),
+    env: { ...process.env, TS_NODE_PROJECT: tsConfig },
+  });
+
+  return (file = 'index.ts', throwErrors = true) =>
+    partialExec(`${command} ${file}`).then((res) => {
+      if (throwErrors && res.err) throw res.err;
+      return res;
+    });
+};
+
+type ModuleType = 'cjs' | 'esm';
+const MODULE_TYPES: ModuleType[] = ['cjs', 'esm'];
+const MODULE_TYPE_ESM = MODULE_TYPES[1];
+
+type ExecBuilderParams = { baseDir: string; command: string };
+const EXEC_BUILDER_PARAMS: Record<ModuleType, ExecBuilderParams> = {
+  cjs: {
+    baseDir: 'cjs-path-mapping',
+    command: CMD_TS_NODE_WITHOUT_PROJECT_FLAG,
+  },
+  esm: {
+    baseDir: 'esm-path-mapping',
+    command: CMD_ESM_LOADER_WITHOUT_PROJECT,
+  },
+};
```

Comment:
To simplify this boilerplate, could probably do

```
EXEC_BUILDER_PARAMS = {...} as const;
```
and then later on
```
for(const [moduleType, execBuilderParams] of Object.entries(EXEC_BUILDER_PARAMS)) {
```

The type system will figure it out.

Reply:
I think `Object.keys` and `Object.entries` have pretty poor type signatures. But as it turns out, I only needed `Object.values`, so this is now looking better :)

I've always used the prefix `<const>` instead of the suffix `as const`; I think it reads more clearly. Any preference?

Reply:
I use the postfix ` as ` syntax for all type assertions since it's compatible with TSX files.  I realize this project doesn't have TSX, but that's just the convention I use everywhere.  So we should stick to that for consistency.